File: server/php/src/Application/Services/PaymentService.php
<?php

namespace OnlinePayments\ExampleApp\Application\Services;

use Exception;
use OnlinePayments\ExampleApp\Application\DTOs\AdditionalPaymentActions\RequestDto;
use OnlinePayments\ExampleApp\Application\DTOs\AdditionalPaymentActions\ResponseDto;
use OnlinePayments\ExampleApp\Application\DTOs\GetPaymentDetails\ResponseDto as GetPaymentDetailsResponseDto;
use OnlinePayments\ExampleApp\Application\DTOs\Payment\RequestDto as CreatePaymentRequestDto;
use OnlinePayments\ExampleApp\Application\DTOs\Payment\ResponseDto as CreatePaymentResponseDto;
use OnlinePayments\ExampleApp\Application\Interfaces\Handlers\PaymentMethodHandlerInterface;
use OnlinePayments\ExampleApp\Application\Interfaces\SDKClients\PaymentClientInterface;
use OnlinePayments\ExampleApp\Application\Interfaces\Services\PaymentServiceInterface;
use OnlinePayments\ExampleApp\Configuration\Exceptions\ValidationException;

readonly class PaymentService implements PaymentServiceInterface
{
    /**
     * @param iterable<PaymentMethodHandlerInterface> $handlers
     */
    public function __construct(
        private PaymentClientInterface $paymentClient,
        private iterable $handlers,
    ) {}

    /**
     * @throws ValidationException
     */
    public function createPayment(CreatePaymentRequestDto $requestDto): CreatePaymentResponseDto
    {
        $handler = null;

        foreach ($this->handlers as $candidate) {
            if ($candidate->getSupportedMethod() === $requestDto->method) {
                $handler = $candidate;
                break;
            }
        }

        if ($handler === null) {
            throw new ValidationException(
                (array)'Unsupported payment method.'
            );
        }

        return $handler->handle($requestDto);
    }

    public function getPaymentDetailsById(string $id): GetPaymentDetailsResponseDto
    {
        return $this->paymentClient->getPaymentDetailsById($id);
    }

    /**
     * @throws Exception
     */
    public function capturePayment(RequestDto $requestDto): ResponseDto
    {
        return $this->paymentClient->capturePayment($requestDto);
    }

    /**
     * @throws Exception
     */
    public function refundPayment(RequestDto $requestDto): ResponseDto
    {
        return $this->paymentClient->refundPayment($requestDto);
    }

    /**
     * @throws Exception
     */
    public function cancelPayment(RequestDto $requestDto): ResponseDto
    {
        return $this->paymentClient->cancelPayment($requestDto);
    }
}

File: server/python/app/application/services/payment_service.py
from app.application.dtos.create_payment.request_dto import CreatePaymentRequestDto
from app.application.dtos.create_payment.response_dto import CreatePaymentResponseDto
from app.application.interfaces.services.payment_service_interface import PaymentServiceInterface
from app.application.exceptions.validation_exception import ValidationException

class PaymentService(PaymentServiceInterface):

    def __init__(self, payment_client, handlers: list):
        self._payment_client = payment_client
        self._handlers = handlers

    def create_payment(self, request_dto: CreatePaymentRequestDto) -> CreatePaymentResponseDto:
        handler = next(
            (h for h in self._handlers if h.supported_method() == request_dto.method),
            None
        )

        if not handler:
            raise ValidationException([f"No handler found for payment method: {request_dto.method}"])
        
        return handler.handle(request_dto)

    def capture_payment(self, request_dto):
        return self._payment_client.capture_payment(request_dto)

    def refund_payment(self, request_dto):
        return self._payment_client.refund_payment(request_dto)

    def cancel_payment(self, request_dto):
        return self._payment_client.cancel_payment(request_dto)

    def get_payment_details(self, id: str):
        return self._payment_client.get_payment_details(id)